pkg/controllers/dashboard/plugin/fscache.go

```diff
@@ -0,0 +1,253 @@
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+	filepathsecure "github.com/cyphar/filepath-securejoin"
+	v1 "github.com/rancher/rancher/pkg/apis/catalog.cattle.io/v1"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	FilesTxtFilename    = "files.txt"
+	PackageJSONFilename = "plugin/package.json"
+
+	// Cache states used by custom resources
+	Cached   = "cached"
+	Disabled = "disabled"
+	Pending  = "pending"
+)
+
+var (
+	FsCache        = FSCache{}
+	FSCacheRootDir = filepath.Join("management-state", "uiplugin")
+	osRemoveAll    = os.RemoveAll
+	osStat         = os.Stat
+	isDirEmpty     = isDirectoryEmpty
+)
+
+type FSCache struct{}
+
+type PackageJSON struct {
+	Version string `json:"version,omitempty"`
+}
+
+// SyncWithControllersCache takes in a slice of UI Plugins objects and syncs the filesystem cache with it
+func (c FSCache) SyncWithControllersCache(cachedPlugins []*v1.UIPlugin) error {
+	for _, p := range cachedPlugins {
+		plugin := p.Spec.Plugin
+		if plugin.NoCache {
+			logrus.Debugf("skipped caching plugin [Name: %s Version: %s] cache is disabled [noCache: %v]", plugin.Name, plugin.Version, plugin.NoCache)
+			continue
+		}
+		if isCached, err := c.isCached(plugin.Name, plugin.Version); err != nil {
+			return err
+		} else if isCached {
+			logrus.Debugf("skipped caching plugin [Name: %s Version: %s] is already cached", plugin.Name, plugin.Version)
+			continue
+		}
+		version, err := getVersionFromPackageJSON(fmt.Sprintf("%s/%s", plugin.Endpoint, PackageJSONFilename))
+		if err != nil {
+			return err
+		}
+		cachedVersion, err := semver.NewVersion(plugin.Version)
+		if err != nil {
+			return err
+		}
+		if !cachedVersion.Equal(version) {
+			return fmt.Errorf("plugin [%s] version [%s] does not match version in controller's cache [%s]", plugin.Name, version.String(), cachedVersion.String())
+		}
+		files, err := readFilesTxt(fmt.Sprintf("%s/%s", plugin.Endpoint, FilesTxtFilename))
+		if err != nil {
+			return err
+		}
+		for _, file := range files {
+			if file == "" {
+				continue
+			}
+			data, err := readFile(plugin.Endpoint + "/" + file)
+			if err != nil {
+				return err
+			}
+			path, err := filepathsecure.SecureJoin(FSCacheRootDir, filepath.Join(plugin.Name, plugin.Version, file))
+			if err != nil {
+				return err
+			}
+			if err := c.Save(data, path); err != nil {
+				logrus.Debugf("failed to cache plugin [Name: %s Version: %s] in filesystem [path: %s]", plugin.Name, plugin.Version, path)
+			}
+		}
+	}
+
+	return nil
+}
+
+// SyncWithIndex syncs up entries in the filesystem cache with the index's entries
+// Entries that aren't in the index, but present in the filesystem cache are deleted
+func (c FSCache) SyncWithIndex(index *SafeIndex, fsCacheFiles []string) error {
+	for _, file := range fsCacheFiles {
+		logrus.Debugf("syncing index with filesystem cache")
+		// Splits /{root}/{pluginName}/{pluginVersion}/* from a fs cache path
+		rel, _ := filepath.Rel(FSCacheRootDir, file)
```

Comment:
We can short-circuit when the file is not relative to the root dir:
```suggestion
		rel, err := filepath.Rel(FSCacheRootDir, file)
		if err != nil {
			continue
		}
```

Reply:
done